src/decoder/stateful.rs

```diff
@@ -0,0 +1,169 @@
+    OutOfResources,
+    #[error("this format is not supported")]
+    UnsupportedFormat,
+    #[error(transparent)]
+    Other(#[from] anyhow::Error),
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum EncodedFormat {
+    AV1,
+    H264,
+    H265,
+    VP8,
+    VP9,
+}
+
+/// Common trait shared by all stateful video decoder backends, providing codec-independent
+/// methods.
+pub trait StatefulDecoderBackend<Codec: StatefulCodec> {
+    /// The type that the backend returns as a result of a decode operation.
+    /// This will usually be some backend-specific type with a resource and a
+    /// resource pool so that said buffer can be reused for another decode
+    /// operation when it goes out of scope.
+    type Handle: DecodedHandle;
+
+    /// Returns the current decoding parameters, as parsed from the stream.
+    fn stream_info(&self) -> Option<&StreamInfo>;
+}
+
+pub trait StatefulCodec {
+    /// State that needs to be kept during a decoding operation, typed by backend.
+    type DecoderState<B: StatefulDecoderBackend<Self>>;
+}
+
+#[derive(Copy, Clone, Debug)]
+
+pub enum ColorGamut {
+    Bt709,
+    Bt470bg,
+    Smpte170m,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum ColorTransfer {
+    Bt709,
+    Smpte170m,
+    Iec61966_2_1,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum ColorMatrix {
+    Rgb,
+    Bt709,
+    Bt470bg,
+    Smpte170m,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct VideoColorSpace {
+    primaries: ColorGamut,
+    transfer: ColorTransfer,
+    matrix: ColorMatrix,
+}
+
+#[derive(Clone, Debug)]
+pub struct EncodedVideoChunk {
+    data: Arc<Vec<u8>>,
+}
+
+pub struct VideoFrame {
+    format: DecodedFormat,
+    coded_resolution: Resolution,
+    display_resolution: Resolution,
+    duration: usize,
+    timestamp: usize,
+    color_space: VideoColorSpace,
+    buffer: Option<Box<dyn AsBytes>>,
+}
+
+pub enum StatefulDecoderEvent {
+    /// The next frame has been decoded.
+    FrameReady(VideoFrame),
+    /// The format of the stream has changed and action is required. A VideoFrame
+    /// with no buffer is returned.
+    FormatChanged(VideoFrame),
+}
+
+pub trait StatefulVideoDecoder<M> {
+    /// Add a new chunk of video to decoding queue, this is a non-blocking method
+    fn decode(&mut self, chunk: EncodedVideoChunk) -> Result<(), DecodeError>;
+
+    /// Return information about the currently decoded stream
+    fn stream_info(&self) -> Option<&StreamInfo>;
+
+    /// Close the decoder and prevent it from future use
+    fn close(&mut self);
+
+    /// Finish processing all pending decode requests
+    fn flush(&mut self) -> Result<(), DecodeError>;
+
+    /// Returns the next event, if there is any pending.
+    fn next_event(&mut self) -> Option<StatefulDecoderEvent>;
+}
+
+#[derive(Copy, Clone, Debug, Default)]
+pub enum StatefulDecoderState {
+    #[default]
+    Unconfigured,
+    Configured,
+    Closed,
+}
+
+pub struct StatefulDecoder<C, B>
```

Comment:
The first use of the stateful decoder interface should be as a stateful wrapper to any stateless decoder, i.e. an adapter that turns a stateless decoder into a stateful one. That adapter maybe be implemented as a `StatefulDecoderBackend`, but for simplicity I'd recommend writing it as an implementation of `StatefulDecoder` first and see what parts we can share with other stateful decoder implementations.

So I'd suggest starting with the design of this wrapper first to guide the interface. Starting from any stateless decoder, the stateful wrapper should

* Accept input buffers and queue them until they can be processed,
* Try to submit work to the stateless decoder, hold if the work cannot be submitted to retry later, and process the events of the stateless decoder, forwarding them to the caller if needed (e.g. decoded frames),

... and I think that's mostly it. The `simple_playback_loop` should give a good example of how things should work, as its purpose is to provide some stateful-ish behavior to our stateless decoders. I will be replaced it by the stateful decoder once it is available.